src/zino/tasks/bgpstatemonitortask.py

```diff
@@ -0,0 +1,404 @@
+import logging
+from dataclasses import dataclass
+from ipaddress import ip_address
+from typing import Any, Iterable, Optional
+
+from pyasn1.type.univ import OctetString
+
+from zino.snmp import NoSuchNameError, SparseWalkResponse
+from zino.statemodels import (
+    BGPAdminStatus,
+    BGPEvent,
+    BGPOperStates,
+    BGPPeerSession,
+    BGPStyle,
+    IPAddress,
+)
+from zino.tasks.task import Task
+
+_logger = logging.getLogger(__name__)
+
+TIME_BEFORE_OPER_DOWN_ALERT = 600
+
+JUNIPER_TRANSLATION_MAP = [
+    ("peer_state", "jnxBgpM2PeerState"),
+    ("peer_admin_status", "jnxBgpM2PeerStatus"),
+    ("peer_remote_address", "jnxBgpM2PeerRemoteAddr"),
+    ("peer_remote_as", "jnxBgpM2PeerRemoteAs"),
+    ("peer_fsm_established_time", "jnxBgpM2PeerFsmEstablishedTime"),
+]
+CISCO_TRANSLATION_MAP = [
+    ("peer_state", "cbgpPeer2State"),
+    ("peer_admin_status", "cbgpPeer2AdminStatus"),
+    ("peer_remote_address", "cbgpPeer2RemoteAddr"),
+    ("peer_remote_as", "cbgpPeer2RemoteAs"),
+    ("peer_fsm_established_time", "cbgpPeer2FsmEstablishedTime"),
+]
+GENERAL_TRANSLATION_MAP = [
+    ("peer_state", "bgpPeerState"),
+    ("peer_admin_status", "bgpPeerAdminStatus"),
+    ("peer_remote_address", "bgpPeerRemoteAddr"),
+    ("peer_remote_as", "bgpPeerRemoteAs"),
+    ("peer_fsm_established_time", "bgpPeerFsmEstablishedTime"),
+]
+LOCAL_AS_OBJECTS = {
+    "juniper": ("BGP4-V2-MIB-JUNIPER", "jnxBgpM2PeerLocalAs"),
+    "cisco": ("CISCO-BGP4-MIB", "cbgpLocalAs"),
+    "general": ("BGP4-MIB", "bgpLocalAs"),
+}
+BUGGY_REMOTE_ADDRESSES = [
+    # Bug in JunOS -- info from IPv6 BGP sessions spill over
+    "0.0.0.0",
+    # Bug in earlier Cisco IOS, info from elsewhere (IPv6?) spills over
+    "32.1.7.0",
+]
+
+
+@dataclass
+class BaseBGPRow:
+    peer_state: BGPOperStates
+    peer_admin_status: BGPAdminStatus
+    peer_remote_address: IPAddress
+    peer_remote_as: int
+    peer_fsm_established_time: int
+
+
+class BGPStateMonitorTask(Task):
+    """Fetches and stores state information about external BGP sessions."""
+
+    async def run(self):
+        bgp_style = await self._get_bgp_style()
+        if bgp_style != self.device_state.bgp_style:
+            _logger.debug(
+                f"Router {self.device_state.name} changed its BGP style from '{self.device_state.bgp_style}' to "
+                f"'{bgp_style}'"
+            )
+            self.device_state.bgp_style = bgp_style
+        if not bgp_style:
+            return
+
+        local_as = await self._get_local_as(bgp_style=bgp_style)
+        uptime = await self._get_uptime()
+        if not local_as or not uptime:
+            return
+
+        if bgp_style == BGPStyle.JUNIPER:
+            bgp_info = await self._get_juniper_bgp_info()
+        elif bgp_style == BGPStyle.CISCO:
+            bgp_info = await self._get_cisco_bgp_info()
+        elif bgp_style == BGPStyle.GENERAL:
+            bgp_info = await self._get_general_bgp_info()
```

Comment:
feels a bit inconsistent that `_get_local_as` can take bgp_style as a parameter while you use if-else statements for deciding which `get_vendor_bgp_info` to use.

It could be useful to make separate classes for handling Juniper/cisco/general that each implement a `get_bgp_info` and `get_local_as`, basically delegating elsewhere the responsibility to decide which style should be used with a factoru pattern or something. The overhead might not be worth it tho.

Reply:
Yes, that idea has also been raised by @hmpf and has been included in #158 for the future.